### PR TITLE
Added Biometric Authentication

### PR DESCRIPTION
**Pull request by akashmeruva9**
Fixes #Issue_Number

<table style="border: 1px solid black;">
            <tr>
                <td  style="border: 1px solid black ;">
                    
                </td>
</table>

https://github.com/user-attachments/assets/aa8aae0c-7051-4e86-8627-9e1e22cc7845

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: app/src/main/java/com/mifos/passcode/PassCodeActivity.kt [hunk 1]
```diff
@@ -29,20 +29,26 @@ class PassCodeActivity : AppCompatActivity() {
...
-            startActivity(Intent(this, LoginActivity::class.java))
-            Toast.makeText(this, "New Screen", Toast.LENGTH_SHORT).show()
-            finish()
+            launchNextActivity()
         }
     }
 
+    private fun launchNextActivity() {
+        startActivity(Intent(this, LoginActivity::class.java))
+        Toast.makeText(this, "New Screen", Toast.LENGTH_SHORT).show()
```
Comment (PratyushSingh07) [inline app/src/main/java/com/mifos/passcode/PassCodeActivity.kt:48]: remove toast

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt [hunk 2]
```diff
@@ -55,7 +67,10 @@ fun PasscodeScreen(
     onSkipButton: () -> Unit,
     onPasscodeConfirm: (String) -> Unit,
     onPasscodeRejected: () -> Unit,
+    onBiometricAuthenticated: () -> Unit,
+    activity: AppCompatActivity
```
Comment (PratyushSingh07) [inline compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:71]: why are we passing it here

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt [hunk 3]
```diff
@@ -64,12 +79,69 @@ fun PasscodeScreen(
...
     var passcodeRejectedDialogVisible by remember { mutableStateOf(false) }
+    val biometricResult by promptManager.promptResults.collectAsState(
+        initial = null
+    )
+    val enrollLauncher = rememberLauncherForActivityResult(
+        contract = ActivityResultContracts.StartActivityForResult(),
+        onResult = {
+            println("Activity result: $it")
+        }
+    )
```
Comment (PratyushSingh07) [inline compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:90]: no need to print this

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt [hunk 4]
```diff
@@ -64,12 +79,69 @@ fun PasscodeScreen(
...
+
+    LaunchedEffect(biometricResult) {
```
Comment (PratyushSingh07) [inline compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:92]: why in launched effect ?

**Review by PratyushSingh07 — CHANGES_REQUESTED** [review]

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt [hunk 5]
```diff
@@ -64,12 +79,69 @@ fun PasscodeScreen(
     val currentPasscode by viewModel.currentPasscodeInput.collectAsStateWithLifecycle()
     val xShake = remember { Animatable(initialValue = 0.0F) }
     var passcodeRejectedDialogVisible by remember { mutableStateOf(false) }
+    val biometricResult by promptManager.promptResults.collectAsState(
```
Comment (PratyushSingh07) [inline compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:82]: collectAsStateWithLifeCycle() in place of collectAsState

**Review by PratyushSingh07 — CHANGES_REQUESTED** [review]

**Comment by PratyushSingh07** [comment]
Also make sure to encrypt the password that the user enters @akashmeruva9

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 4 above)
Comment (akashmeruva9) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:92]: Here we are observing the biometric result, if the biometric result = Authentication.NotSet, then it launches the activity to enroll the biometrics.
We can use an if statement here, but this gets triggered on every recomposition.

**Review by PratyushSingh07 — CHANGES_REQUESTED** [review]

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 4 above)
Comment (PratyushSingh07) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:92]: what are the cases in which we can Authentication.NotSet as a result of `biometricResult`

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 5 above)
Comment (akashmeruva9) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:82]: here biometric result is not a StateFLow intitially , so we have to initialize it with null here.

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 4 above)
Comment (akashmeruva9) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:92]: when this happens : 

BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
    resultChannel.trySend(BiometricResult.AuthenticationNotSet)
    return
}

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 2 above)
Comment (PratyushSingh07) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:71]: use context or this to pass reference to an activity.  what you have done is not the solution

File: compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt (hunk 4 above)
Comment (akashmeruva9) [reply compose/src/main/java/com/mifos/compose/component/PassCodeScreen.kt:92]: when we launch this
  
```
    LaunchedEffect(true){
        if(preferenceManager.hasPasscode) {
            promptManager.showBiometricPrompt(
                title = context.getString(R.string.biometric_auth_title),
                description = context.getString(R.string.biometric_auth_description)
            )
        }
    }
```


the biometric manager checks for fifferent conditions, 

```
    fun showBiometricPrompt(
        title: String,
        description: String
    ) {
        val manager = BiometricManager.from(activity)
        val authenticators = if(Build.VERSION.SDK_INT >= 30) {
            BIOMETRIC_STRONG or DEVICE_CREDENTIAL
        } else BIOMETRIC_STRONG

        val promptInfo = PromptInfo.Builder()
            .setTitle(title)
            .setDescription(description)
            .setAllowedAuthenticators(authenticators)

        if(Build.VERSION.SDK_INT < 30) {
            promptInfo.setNegativeButtonText("Cancel")
        }

        when(manager.canAuthenticate(authenticators)) {
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> {
                resultChannel.trySend(BiometricResult.HardwareUnavailable)
                return
            }
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> {
                resultChannel.trySend(BiometricResult.FeatureUnavailable)
                return
            }
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
                resultChannel.trySend(BiometricResult.AuthenticationNotSet)
                return
            }
            else -> Unit
        }
```

then when this happens it updates the biometric result

```
BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> {
resultChannel.trySend(BiometricResult.AuthenticationNotSet)
return
}
```

**Pull request closed without merging**